### PR TITLE
fix(hmr): improve detecting of ReactRefreshBoundary and Return value of `React.createContext` is not considered as React component

### PR DESCRIPTION
**Pull request by underfin**
## Summary

Thanks for @hyf0 to resolve the issue with me.

- fix implement of react referesh, port from `react-refresh-webpack-plugin`,it should detect component at runtime, it can fix react context hmr issue
- implement `__webpack_module__.id`

## Related issue (if exists)
close https://github.com/web-infra-dev/rspack/issues/2266

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [ ] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [ ] I have added changeset via `pnpm run changeset`.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by ahabhgk — DISMISSED** [review]

**Comment by hyf0** [comment]
Could you link the issues this PR fixed?

**Review by hyf0 — APPROVED** [review]

**Review by hyf0 — DISMISSED** [review]

**Review by hyf0 — DISMISSED** [review]

**Review by hardfist — DISMISSED** [review]

**Review by bvanjoi — DISMISSED** [review]

File: packages/rspack-dev-client/src/reactRefresh.ts [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 RefreshRuntime.injectIntoGlobalHook(globalThis);
-globalThis.$RefreshReg$ = () => {};
-globalThis.$RefreshSig$ = () => type => type;
 
-var queueUpdate = debounce(RefreshRuntime.performReactRefresh, 16);
+// Port from https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/loader/utils/getRefreshModuleRuntime.js#L29
+function refresh(moduleId, webpackHot) {
+	const currentExports = RefreshUtils.getModuleExports(moduleId);
+	const fn = () => {
+		RefreshUtils.executeRuntime(currentExports, moduleId, webpackHot);
```
Comment (h-a-n-a) [inline packages/rspack-dev-client/src/reactRefresh.ts:12]: Is this right for promise exported value?

Comment (underfin) [reply packages/rspack-dev-client/src/reactRefresh.ts:12]: Yes

File: packages/rspack-dev-client/src/reactRefresh.ts [hunk 2]
```diff
@@ -1,22 +1,26 @@
...
-
 RefreshRuntime.injectIntoGlobalHook(globalThis);
-globalThis.$RefreshReg$ = () => {};
-globalThis.$RefreshSig$ = () => type => type;
 
-var queueUpdate = debounce(RefreshRuntime.performReactRefresh, 16);
+// Port from https://github.com/pmmmwh/react-refresh-webpack-plugin/blob/main/loader/utils/getRefreshModuleRuntime.js#L29
+function refresh(moduleId, webpackHot) {
+	const currentExports = RefreshUtils.getModuleExports(moduleId);
+	const fn = () => {
```
Comment (h-a-n-a) [inline packages/rspack-dev-client/src/reactRefresh.ts:11]: It seems the function should receive the exported value:
image

**Review by h-a-n-a — APPROVED** [review]

**Review by ahabhgk — APPROVED** [review]

**Review by hardfist — APPROVED** [review]

**Pull request merged**